Add deployment for host package

backspace opened the pull request:
This is largely adapted from safe-tools-client in cardstack/cardstack. It adds:
* a workflow that will allow Cardie to deploy the application
* a job in the existing `ci.yaml` that deploys to staging upon a successful test on `main`
  * [here](https://boxel-host-staging.stack.cards/), generated before adding the `if` that restricts the branch
* preview deployments for PRs
  * [staging](https://deployment.boxel-host-preview.stack.cards/)
  * [production](https://deployment.boxel-host-preview.cardstack.com/)

Some assumptions:
* the application is built with `DEMO_REALM_URL`, pending the merging of #177
* the realm server domains are `realm-demo-staging.stack.cards` and `realm-demo.cardstack.com`
  * these do not yet exist, that work is forthcoming

backspace committed:
Add deployment for host package

This is largely adapted from safe-tools-client in
cardstack/cardstack, which is the Ember application
that’s had deployment added most recently. It will
surely take repeated fixes.

backspace committed:
Add missing version config

backspace committed:
Change manual deployment filename for now

backspace committed:
Add hack to force previews

backspace committed:
Add shell for script

whyeeee

backspace committed:
Fix deployment script invocation

backspace committed:
Add prefix for ember command

backspace committed:
Add Node memory override

backspace committed:
Add workflow with expected name

backspace committed:
Remove misnamed file

The true problem here is that Cardie can’t deploy from
workflow on a branch… of course

backspace committed:
Add temporary workflow to always deploy

backspace committed:
Remove env placeholder

backspace committed:
Fix indentation

YAML!!!

backspace committed:
Add shell specification

Why does this work elsewhere without it?

backspace committed:
Change use action from monorepo

Will it work? 🤔

backspace committed:
Add branch

backspace committed:
Add full path

Why is the documentation so tragique

backspace committed:
Change version references to hardcoded for now

backspace committed:
Add Node option override again

backspace committed:
Change to use pnpm exec instead of modules .bin

backspace committed:
Remove note

backspace committed:
Remove unused env

backspace committed:
Add config of realm server environment variable

backspace committed:
Change hardcoded values to inputs with defaults

env wasn’t working because this is an action, not a
workflow.

backspace committed:
Move memory override into scripts

backspace committed:
Move environment variables

backspace committed:
Remove note

backspace committed:
Add realm server URL for preview deployments

backspace committed:
Add step for deployment on main

backspace committed:
Change to same version of AWS credentials action

Why is it working in pr-boxel-host.yml but not here?

backspace committed:
Add permissions

As suggested here:
https://github.com/aws-actions/configure-aws-credentials/issues/271#issuecomment-931012696

backspace committed:
Change ci.yaml deployment to only happen on main

backspace committed:
Fix change detection for previews

habdelra commented:
based on your assumptions it sounds like we have only one server to host both the base realm and demo realm for each environment (which is fine), but it means that we'll need to land https://linear.app/cardstack/issue/CS-5132/start-realms-in-series first. (i plan on tackling that today).

habdelra commented on package.json:
```diff
@@ -5,6 +5,9 @@
     "url": "https://github.com/cardstack/boxel-motion"
   },
   "scripts": {
+    "deploy:boxel-host": "cd packages/host && NODE_OPTIONS='--max_old_space_size=4096' pnpm exec ember deploy",
```
I think (?) the max_old_space_size is only necessary when building for a fastboot server too. we can use `FASTBOOT_DISABLED=true` env var to not do the fastboot build, which i think means you wont have to increase the heap size. not doing the fastboot build should still be ok for the card prerendering needs of the realm server. it just needs the conventional build's dist artifacts to work.

backspace replied:
I’ll try that, thanks

backspace replied:
The heap error [still happened](https://github.com/cardstack/boxel-motion/actions/runs/3951287150/jobs/6764949598#step:4:386)!

habdelra replied:
argh, i guess the heap error is not tied to the fastboot build, but rather its tied to performing a production build. sorry about that. i guess we need to keep the max_old_space_size in there. You can still leave in `FASTBOOT_DISABLED=true`, though, as that will result in a quicker build.

habdelra replied:
or wait, i'm sorry, i think it will actually do nothing since we are doing a build only and not serving. (i'm getting building conflated with serving)

habdelra reviewed (approved):
looks good

backspace commented:
> based on your assumptions it sounds like we have only one server to host both the base realm and demo realm for each environment (which is fine), but it means that we'll need to land https://linear.app/cardstack/issue/CS-5132/start-realms-in-series first. (i plan on tackling that today).

I failed to revisit this after I understood that there are two realm servers. My current understanding is that demo realm server consumes base, so host should point to demo? I can change it today to point to the prospective demo URL.

backspace committed:
Merge branch 'main' into deployment

# Conflicts:
#	packages/host/package.json
#	pnpm-lock.yaml

backspace committed:
Change builds to skip Fastboot

As suggested by @habdelra:
https://github.com/cardstack/boxel-motion/pull/178/files#r1073787932

backspace committed:
Change realm URLs

habdelra commented:
Confirm, the host should point to the demo realm. 

However if we do decide that we want both the base and demo realms on the same server, i just posted the PR to start realms on a single server serially (https://github.com/cardstack/boxel-motion/pull/185). I personally don't have a preference.

backspace committed:
Revert "Change builds to skip Fastboot"

This reverts commit 7cd261144e920946fe1e50ace23b071c2fe3e64b.
